### PR TITLE
Add asset files for Linux desktop

### PR DESCRIPTION
**Pull request by srakitnican**
Issue #58.

.desktop file slightly modified from original:
- Removed Encoding since it causes a deprecated warning with desktop-file-utils
- Updated Comment based on the same from appdata

### CODE REVIEW

**Comment by piernov** [comment]
Gonna have to try that but it looks great. Travis is (hopefully) currently generating the packages on my repo.

**Comment by MattSturgeon** [comment]
Looking at the changes, this might also fix #33

**Comment by piernov** [comment]
It does but only for GNU/Linux. On Windows we still have to make changes to the registry, and for macOS I don't have a clue.

**Comment by piernov** [comment]
CMake-generated RPM and DEB packages available here: https://github.com/piernov/OpenBoardView/releases/tag/N13 . I will take a look at them tomorrow.

**Comment by MattSturgeon** [comment]
Regarding  `make install`s, I think the install target should always act like a full system install, it should be agnostic of the install location (`/`, package-dest or some local directory).

That said you can always add cmake CACHE variables to configure the build, like how within the install location you can also set the prefix (and on some build systems set separate prefixes for different file types, e.g. `/usr/local` for bins and `/usr` for `.desktop`).

There's nothing stopping you making a toggle for "install asset files", though realistically it's probably best to just include them in all installs, incase someone wants to use them for something manually.

**Comment by piernov** [comment]
Yep good point. So it's fine like that. We only have to make sure everthing works as it should.

File: asset/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,15 @@ if(APPLE)
...
 	)
+elseif(UNIX)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/ofbv-app.svg
+		DESTINATION share/icons/hicolor/scalable/apps RENAME openboardview.svg)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/openboardview.appdata.xml
+		DESTINATION share/appdata)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/openboardview.desktop
+		DESTINATION share/applications)
+	install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/openboardview.xml
+		DESTINATION share/mime/packages)
```
Comment (MattSturgeon) [inline asset/CMakeLists.txt]: I think the paths should be `/usr/share/` not `share/`. CMake will prefix relative paths with `CMAKE_INSTALL_PREFIX` (usually `/usr/local` or `/usr`), but these files should always go under `/usr`.

Comment (srakitnican) [inline asset/CMakeLists.txt]: Well if you are installing under `/usr/local`, you would want to install the files under `/usr/local/share`, as well?

Comment (srakitnican) [inline asset/CMakeLists.txt]: https://specifications.freedesktop.org/shared-mime-info-spec/shared-mime-info-spec-latest.html#s2_layout

Comment (srakitnican) [inline asset/CMakeLists.txt]: This also makes it possible to install OBV in in $HOME directory for people that can't write to system directories, for example:
make install DESTDIR=$HOME/.local

Comment (MattSturgeon) [inline asset/CMakeLists.txt]: Agreed. Do you have a PR for srakitnican/OpenBoardView@8e5aaac (and any other useful changes)?

Probably should have a separate PR with that change and to refactor `build.sh` slightly to make use of DESTDIR

**Comment by piernov** [comment]
Also for MIME types we have (somewhat working) CAST support (see #59 ).

The other MIME types could be a bit more descriptive. For example `application/x-openboardview-brd-encoded` could be `application/x-openboardview-brd-landrex` or `application/x-openboardview-brd-testlink` or whatever. Although it's still OK like as it is now.

Maybe you could rename the `asset/ofbv-app.svg` to `asset/openboardview.svg` directly in the Git repo? (the `ofbv` part doesn't make sense anymore anyway)

**Comment by MattSturgeon** [comment]
Re the icon name: probably better to do that separately, perhaps a PR to review asset names in general (e.g. should the other icons still be named `brd`)?

I agree it should be renamed, and itd be nice to have it done before this, but it feels kinda out of scope here, especially if we are going to squash merge.

More descriptive MIME names would be nice, realistically if it's including the program name, does it need the extension too (e.g. `application/x-landrex` or `application/x-openboardview-landrex`)? We should probably check any of these filetypes already have "official" mimetypes too.

**Comment by piernov** [comment]
Ok, I was talking about `ofbv-app.png` in particular because @srakitnican explicitely renames it when installing the asset file.

About the MIME type names I've no idea. Maybe we shouldn't bother too much and leave it at that.

**Comment by MattSturgeon** [comment]
Probably best if @inflex and @piernov take a look at the "magic" MIME detection values just to ensure they will work well for all filetypes.

It would be a shame if we check against a string or a byte that isn't always the same (or give false positives from other filetypes)

**Comment by piernov** [comment]
Should probably be matching with https://github.com/OpenBoardView/OpenBoardView/blob/master/src/openboardview/BoardView.cpp#L422 (and the various verifyFormat() methods).
I think I was too lazy with the CST file…

**Comment by srakitnican** [comment]
Collisions are not fatal, this is just a convenience thing, for example if other application claims the same extension with greater weight it can have it, but if matches any magic bits, it should be preferred by OBV. User also gets a nice recommended applications overview in "Open With" menu.

Why `application/x-openboardview-...` so that you know what to erase when it gets all over your files :) There is nowhere to be found specifications for any of these file types, nor what it is called. So I thought the best here is to name it after application that installed this extension. But feel free to change it later to any of the likings.

I am not very familiar with all the file types, so I wrote to the best I could based on source code found here and information found mostly from your posts.

**Comment by MattSturgeon** [comment]
### Overview of filetype matching done by BoardView.cpp and the File classes.

#### using file extension only

FZFile ".fz"
ASCFile ".bom" ".asc"
CSTFile ".cst"

#### Comparing the buffer as hex

BRDFile
  - Check first 32 bits match `{{0x23, 0xe2, 0x63, 0x28}}`

#### using a string search

BRDFile2
  - `find_str("BRDOUT:") && find_str("NETS:")`

BDVFile
  - `find_str("dd:1.3?,r?-=bb")`
    - or
  - `find_str("<<format.asc>>") && find_str("<<pins.asc>>")`

BVRFile
  - `find_str("BVRAW_FORMAT_1")`

#### For comparison, here's how it is checked in mime-info

x-openboardview-board
  - filename "*.{bdv,bvr,fz}"
    - @srakitnican is `{,}` globbing supported for mimetypes? If so it would compact three lines into one.

x-openboardview-asc
  - filename "*.asc"
  - position 8:60 matches "eM-Test Expert"

x-openboardview-brd-landrex
  - filename "*.brd"
  - Check the first 32 bits match `0x23e26328` _(same as code, but with addition of file extension)

x-openboardview-brd-r4
  - filename "*.brd"
  - position 0:30 matches "BRDOUT:"

x-openboardview-lst
  - filename "*.lst"
  - position 3:120 matches "$ASM"

### Thoughts

I notice one of the main differences is that the magic mine checks an exact row:column position, while the C++ implementation does a full string search. While I'm rather unfamiliar with the file formats, it feels like the precision of the row:column check could be prone to breaking for some of the cases.

Also worth checking if there's any filetypes that could be checked better (either here or in our code) based on the other implementation. Again, just raising potential points, not my area of expertise.

---

If all of the above is ok _(or once it has been used for improvements)_ I think this may be ready for a merge.

**Comment by srakitnican** [comment]
> x-openboardview-board
> 
> filename "*.{bdv,bvr,fz}"
> @srakitnican is {,} globbing supported for mimetypes? If so it would compact three lines into one.

I don't think so, at least I haven't seen anyone using it. In fact the only thing I saw is `*`.

> x-openboardview-lst
>
> filename "*.lst"
> position 3:120 matches "$ASM"

If position 0 matches "$VS2"

**Comment by MattSturgeon** [comment]
> > x-openboardview-lst
> > 
> > filename "*.lst"
> > position 3:120 matches "$ASM"
>
> If position 0 matches "$VS2"

Oops, edited.

> the only thing I saw is `*`.

Yeah, `{ }` list globs aren't used/known as much, but that doesn't mean they aren't supported. Not a big deal either way.

---

So are you happy with how the file matching compares between here and obv code? Are there any magic checks that could go? _(filename globs are prefered for performance reasons because the system doesn't have to open the file to do the check)._ Are there any checks that should be added? Should any magic checks by nested under the glob check or vice versa? Are there any checks that are too vague and might match other files we can't read? If so is that an issue?

**Comment by srakitnican** [comment]
Well I have tested only on files attached in issues, and manually generated files of scenarious I can think of. What is there in the wild I have no clue. As I said already it would help tremendously is we would work by some specification, as we can't do that the only thing left is to do it by trial and error. I can't think any reason where false matching or non matching would cause any major issues except annoyance.

> ASCFile ".bom" ".asc"

I didn't add .bom as I've read is not supported currently.

> BRDFile2
> find_str("BRDOUT:") && find_str("NETS:")

Searching only for "BRDOUT:", less searching the better as I have no firm indication of where other string could be, and searching for it in entire file is not a good idea as I've read.

> BDVFile
>
> find_str("dd:1.3?,r?-=bb")

I've missed this one, if you think is neccessary and consistent I could add it

> or
> find_str("<<format.asc>>") && find_str("<<pins.asc>>")

I don't think I can search for this one because file is encoded originally as I understand.

> BVRFile
> find_str("BVRAW_FORMAT_1")

If you think is neccesary I could add this one as well, although searching for it in entire file is not a good idea, if you know for sure of the location(s).

**Comment by MattSturgeon** [comment]
Searches are more costly than looking at a specific case, however they shouldn't be much of an issue if they only happen after checking something cheaper (e.g. nested under the glob).

Still, it's minor nit-picking, we can always open issues in future to improve the file matching both for the mime types and in the code.

@piernov are you happy to squash and merge?

**Comment by srakitnican** [comment]
> Searches are more costly than looking at a specific case, however they shouldn't be much of an issue if they only happen after checking something cheaper (e.g. nested under the glob).

Except they are not nested, try removing the extension, they are still searched for and matched :)

**Comment by MattSturgeon** [comment]
I wasn't saying it _is_ nested, I was saying it possibly could be - if we modified the XML structure.

For example for `"BRDOUT:" && "NETS:"`, you could nest `"NETS:"` under your existing `"BRDOUT"`, but it isn't really needed.

Either way it isn't important ATM, as I said in my last comment, assuming @piernov is happy I think this is probably ready for merge.

**Comment by srakitnican** [comment]
Nesting `<match>` is possible yes, but nesting `<match>` under `<glob>` is not. Yes your point about nesting `NETS:` under `BRDOUT:` shouldn't be too expensive, I guess we could do that if we are sure we can find it there.

**Comment by MattSturgeon** [comment]
> I notice one of the main differences is that the magic mine checks an exact row:column position, while the C++ implementation does a full string search.

Doh! Just read [the spec](https://specifications.freedesktop.org/shared-mime-info-spec/shared-mime-info-spec-0.11.html#idm139839923550176) and it turns out I'm an idiot, the `:` is indicating a range of bytes to search, not an exact `row:column` to check.

I think I'll just merge this since I can think of no reason why it will cause any issues.

**Comment by piernov** [comment]
Seems to work with various FZ/BRD/BDV/ASC (what I've got here right now) so I'm happy with that.

**Comment by srakitnican** [comment]
What we could do for BRD file type if we don't want to associate with any other files is to remove extension association and rely only on magic. That way it will not try to open Cadence file format for example.

**Pull request merged**